geo/src/algorithm/validation/mod.rs

```diff
@@ -0,0 +1,154 @@
+pub use multi_polygon::InvalidMultiPolygon;
+pub use point::InvalidPoint;
+pub use polygon::InvalidPolygon;
+pub use rect::InvalidRect;
+pub use triangle::InvalidTriangle;
+
+use std::boxed::Box;
+use std::fmt;
+
+/// A trait to check if a geometry is valid and report the reason(s) of invalidity.
+///
+/// ```
+/// use geo::algorithm::Validation;
+/// use geo::wkt;
+///
+/// let valid_polygon = wkt!(POLYGON((0. 0., 1. 1., 1. 0., 0. 0.)));
+/// assert!(valid_polygon.is_valid());
+///
+/// let invalid_polygon = wkt!(POLYGON((0. 0., 1. 1.),(3. 3., 3. 4.,4. 4.)));
+/// assert!(!invalid_polygon.is_valid());
+///
+/// // Get the first validation error, as a `Result`
+/// let validation_error = invalid_polygon.check_validation().unwrap_err();
+/// use geo::algorithm::validation::{InvalidPolygon, RingRole};
+/// assert_eq!(validation_error, InvalidPolygon::TooFewPointsInRing(RingRole::Exterior));
+///
+/// // Get a human readable error
+/// let text = validation_error.to_string();
+/// assert_eq!(text, "exterior ring must have at least 3 distinct points");
+//
+/// // Get all validation errors
+/// let all_validation_errors = invalid_polygon.validation_errors();
+/// assert_eq!(all_validation_errors.len(), 2);
+/// assert_eq!(all_validation_errors[0].to_string(), "exterior ring must have at least 3 distinct points");
+/// assert_eq!(all_validation_errors[1].to_string(), "interior ring at index 0 is not contained within the polygon's exterior");
+/// ```
+pub trait Validation {
+    type Error: std::error::Error;
+
+    /// Check if the geometry is valid.
+    fn is_valid(&self) -> bool {
+        self.check_validation().is_ok()
+    }
+
+    /// Return the reason(s) of invalidity of the geometry.
+    ///
+    /// Though we try to return *all* problems with a geometry, it's possible that previous errors
+    /// will obscure subsequent errors. For example, a MultiPolygon requires all its elements to be
+    /// valid and non-overlapping. If one of the individual polygons is invalid, we can't guarantee
+    /// the correctness of their "overlap" check which assumes valid input. Therefore, you should
+    /// re-validate after attempting to correct any validation errors.
+    fn validation_errors(&self) -> Vec<Self::Error> {
+        let mut validation_errors = Vec::new();
+
+        self.visit_validation(Box::new(|problem| {
+            validation_errors.push(problem);
+            Ok::<(), Self::Error>(())
+        }))
+        .expect("no errors are returned");
+
+        validation_errors
+    }
+
+    /// Return the first reason of invalidity of the geometry.
+    fn check_validation(&self) -> Result<(), Self::Error> {
+        self.visit_validation(Box::new(Err))
+    }
+
+    /// Visit the validation of the geometry.
+    ///
+    /// The closure `handle_validation_error` is called for each validation error.
+    fn visit_validation<T>(
+        &self,
+        handle_validation_error: Box<dyn FnMut(Self::Error) -> Result<(), T> + '_>,
```

Comment:
Presumably something like this wasn't possible?

```
handle_validation_error: FnMut(Self::Error) -> Result<(), T>,
```

Reply:
Because of the potentially infinite recursion of `GeometryCollection`, we cannot use a true generic `impl FnMut` - it must be boxed. 

Is that what you were asking about?

Reply:
Here's the error message when you switch to: `handle_validation_error: impl FnMut(Self::Error) -> Result<(), T>,`

```
error: reached the recursion limit while instantiating `validation::geometry_collection::<impl Validation for geo::GeometryCollection>::visit_validation::<geo::validation::InvalidGeometry, Box<{closure@validation::geometry::<impl Validation for geo::Geometry>::visit_validation<..., ...>::{closure#7}}>>`
  --> /Users/mkirk/src/georust/geo/geo/src/algorithm/validation/geometry.rs:91:48
   |
91 |               Geometry::GeometryCollection(g) => g.visit_validation(Box::new(|err| {
   |  ________________________________________________^
92 | |                 handle_validation_error(InvalidGeometry::InvalidGeometryCollection(err))
93 | |             }))?,
   | |_______________^
   |
note: `validation::geometry_collection::<impl Validation for geo::GeometryCollection<F>>::visit_validation` defined here
  --> /Users/mkirk/src/georust/geo/geo/src/algorithm/validation/geometry_collection.rs:28:5
   |
28 | /     fn visit_validation<T>(
29 | |         &self,
30 | |         mut handle_validation_error: impl FnMut(Self::Error) -> Result<(), T>,
31 | |     ) -> Result<(), T> {
   | |______________________^
   = note: the full type name has been written to '/Users/mkirk/src/georust/geo/target/debug/deps/jts_test_runner-201a42bc420bd6fb.long-type.txt'


```

Reply:
I don't think that addresses the potentially infinite recursion of the closure - without boxing the closure, the compiler must build a type which can store the state of a closure which validates a `GeometryCollection` inside a `GeometryCollection` inside a `GeometryCollection` inside a...

I confess I'm not that confident in my analysis, but I did have a go at your suggestion [here](https://github.com/georust/geo/tree/mkirk/mthh_is_valid_2-code-review-closure-unboxing), but ultimately hit the same compiler error:

> Reached the recursion limit while instantiating `validation::geometry_collection::<impl Validation for geo::GeometryCollection>::visit_validation::<geo::validation::InvalidGeometry, {closure@validation::geometry::<impl Validation for geo::Geometry>::visit_validation<geo::validation::InvalidGeometry, ...>::{closure#7}}>`